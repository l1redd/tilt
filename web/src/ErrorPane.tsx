import React, { PureComponent } from "react"

/*
  ErrorPane is a list of error items
    - You can click to expand(??)
    - Each resource has its own error pane
    - You can click from an error item to a resource log

 1. Build errors
    - Build in BuildHistory with an error, grab the entire log

 2. Runtime errors
 3. Tiltfile errors
*/

type Build = {
  log: string
  finishTime: string
  error: {} | null
}

type ErrorResource = {
  name: string
  buildHistory: Array<Build>
}

type ErrorsProps = {
  resources: Array<ErrorResource>
}

class ErrorPane extends PureComponent<ErrorsProps> {
  render() {
    let errorElements: Array<JSX.Element> = []
    this.props.resources.forEach(r => {
      r.buildHistory.forEach((b, i) => {
        if (b.error !== null) {
          errorElements.push(<li key={r.name + i}>{b.log}</li>)
        }
      })
    })

    if (errorElements.length === 0) {
      return <p>No errors</p>
    }

    return <ul>{errorElements}</ul>
  }
}

export default ErrorPane
