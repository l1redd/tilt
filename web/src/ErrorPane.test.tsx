import React from "react"
import ReactDOM from "react-dom"
import ErrorPane from "./ErrorPane"
import renderer from "react-test-renderer"

it("renders no errors", () => {
  let resources = [
    {
      name: "foo",
      buildHistory: [],
    },
  ]

  const tree = renderer.create(<ErrorPane resources={resources} />).toJSON()

  expect(tree).toMatchSnapshot()
})

it("renders one error", () => {
  const ts = "1,555,970,585,039"
  let resources = [
    {
      name: "foo",
      buildHistory: [
        {
          log: "laa dee daa I'm an error",
          finishTime: ts,
          error: {},
        },
      ],
    },
  ]

  const tree = renderer.create(<ErrorPane resources={resources} />).toJSON()

  expect(tree).toMatchSnapshot()
})

it("renders one resource with two build errors", () => {
  const ts = "1,555,970,585,039"
  let resources = [
    {
      name: "foo",
      buildHistory: [
        {
          log: "laa dee daa I'm an error",
          finishTime: ts,
          error: {},
        },
        {
          log: "laa dee daa I'm another error",
          finishTime: ts,
          error: {},
        },
      ],
    },
  ]

  const tree = renderer.create(<ErrorPane resources={resources} />).toJSON()

  expect(tree).toMatchSnapshot()
})
