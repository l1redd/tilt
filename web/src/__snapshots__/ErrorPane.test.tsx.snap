// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`renders no errors 1`] = `
<p>
  No errors
</p>
`;

exports[`renders one error 1`] = `
<ul>
  <li>
    laa dee daa I'm an error
  </li>
</ul>
`;

exports[`renders one resource with two build errors 1`] = `
<ul>
  <li>
    laa dee daa I'm an error
  </li>
  <li>
    laa dee daa I'm another error
  </li>
</ul>
`;
